server/liveblog/advertisements/amp/utils.py

```diff
@@ -0,0 +1,172 @@
+
+
+def inject_advertisments(content, settings, ads_list, theme):
+    """
+    This method receives a parsed with BeautifulSoup content and injects the
+    advertisments according to given `settings` (frequency, order, etc)
+
+    Raises `ParameterError` if parameters do not match the require type
+
+    Note:
+        At the moment advertisements injection has been only tested on Default AMP Theme
+
+    Args:
+        content (:obj:`bs4.BeautifulSoup`): BeautifulSoup parsed content
+        settings (:obj:`AdsSettings`): Required settings to query and injects content into soup
+        ads_list (`list`): List of advertisments entries
+        theme: (:obj: `liveblog.themes.ThemesResource` instance): Blog theme
+    """
+
+    if not isinstance(content, bs4.BeautifulSoup):
+        raise ParameterError(
+            "Wrong `content` parameter type. Expected `bs4.BeautifulSoup` instance")
+
+    if not isinstance(settings, AdsSettings):
+        raise ParameterError(
+            "Wrong `settings` parameter type. Expected `AdsSettings` instance")
+
+    def _not_tombstone_item(css_class):
+        return settings.article_class in css_class and settings.tombstone_class not in css_class
+
+    articles = content.find_all(
+        lambda x: x.name == settings.article_tag and _not_tombstone_item(x.get('class', [])))
+
+    pcount = len(articles)
+    acount = len(ads_list)
+    frequency = settings.frequency
+    template = settings.template
+
+    if settings.order == DESC:
+        ads_list = ads_list[::-1]
+
+    theme_service = get_resource_service('themes')
+    theme_settings = theme_service.get_default_settings(theme)
+
+    for i in range(0, pcount, frequency):
+        index = math.ceil(i / frequency) % acount
+
+        ref_article = articles[i]
+        item = ads_list[index]
+        ad_id = "ads_%s_%s" % (item['_id'], i)
+
+        ad_content = template.render(
+            ad_id=ad_id, item=item, theme=theme, settings=theme_settings)
+        ad_content = prepare_amp_content(ad_content)
+
+        # we need to increment the data-update-time in order to trigger amp updating
+        ref_article['data-update-time'] = int(ref_article['data-update-time']) + 1
+        ref_article.append(ad_content)
+
+
+def prepare_amp_content(content):
+    """
+    This ensure that ad html rendered content to be AMP compatible
+
+    Args:
+        content (`str`): HTML string content
+
+    Returns:
+        `BeautifulSoup` adjusted content to be compatible with AMP
+    """
+
+    # using html.parser to avoid html and body tags
+    ad_content = bs4.BeautifulSoup(content, "html.parser")
+
+    # time to adjust couple of things from advertisement render
+    # e.g: image and some wrong attributes
+    ad_image = ad_content.find('img')
+    if (ad_image):
+        amp_img = ad_content.new_tag('amp-img')
+        amp_img['src'] = ad_image.attrs['src']
+        amp_img['width'], amp_img['height'] = 600, 200
```

Comment:
We have on image meta size information if needed

Reply:
and... I checked `_meta` and it doesn't seem to have image size information

```javascript
'meta': {
  'data': {
    'text': 'This is Steve ;)', 
    'image': {
      'picture_url': 'http://localhost:5000/api/upload/5ab27f7ce83729cd71d16dc5/raw?_schema=http', 
      'picture': 'urn:newsml:localhost:2018-03-21T09:51:24.808352:0c0ef8b9-23d2-4800-9b0a-54def278a641'
    }, 
    'link': ''
  }
}
```